### PR TITLE
Order Quantity Type + Amount Change Crash Fix

### PR DESCRIPTION
**Pull request by falnatsheh**
- https://www.pivotaltracker.com/story/show/158305029
- https://www.pivotaltracker.com/story/show/158329465
- https://www.pivotaltracker.com/story/show/157754041

### CODE REVIEW

File: TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift [hunk 1]
```diff
@@ -416,8 +416,25 @@ class TradeItYahooEquityTradingTicketViewController: TradeItYahooViewController,
...
+                    code: TradeItErrorCode.systemError
+                )
+                self.alertManager.showError(errorResult, onViewController: self)
+                return cell
+            }
+            
+            self.order.quantityType = supportedOrderQuantityTypes.contains(self.order.quantityType) ? self.order.quantityType : supportedOrderQuantityTypes[0]
             let quantitySymbol = self.order.quantityType == OrderQuantityType.shares ? "Shares" : self.order.linkedBrokerAccount?.accountBaseCurrency
+            
+            let cell = cell as? TradeItNumericToggleInputCell
```
Comment (falnatsheh) [inline TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift]: I moved this down so I could use the previous none optional `cell` as return if an error occurred.

File: TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift [hunk 2]
```diff
@@ -416,8 +416,25 @@ class TradeItYahooEquityTradingTicketViewController: TradeItYahooViewController,
...
+            let supportedOrderQuantityTypes = orderCapabilities.supportedOrderQuantityTypes(
+                forAction: self.order.action,
+                priceType: self.order.type
+            )
+            
+            if supportedOrderQuantityTypes.isEmpty {
+                let errorResult = TradeItErrorResult(
+                    title: "Order Quantity Type Error",
+                    message: "No valid quantity type found for this action",
+                    code: TradeItErrorCode.systemError
```
Comment (falnatsheh) [inline TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift]: Not sure what error code this fits under, this seems the closest but let me know if you I need to change it.

File: TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift [hunk 3]
```diff
@@ -416,12 +416,29 @@ class TradeItYahooEquityTradingTicketViewController: TradeItYahooViewController,
...
+            
+            self.order.quantityType = supportedOrderQuantityTypes.contains(self.order.quantityType) ? self.order.quantityType : supportedOrderQuantityTypes[0]
             let quantitySymbol = self.order.quantityType == OrderQuantityType.shares ? "Shares" : self.order.linkedBrokerAccount?.accountBaseCurrency
+            
+            let cell = cell as? TradeItNumericToggleInputCell
             cell?.configure(
                 onValueUpdated: { newValue in
                     self.order.quantity = newValue
-                    self.reload(row: .estimatedChange)
+                    self.updateEstimatedChangeRow()
```
Comment (falnatsheh) [inline TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift]: Below are the changes to fix the crash when changing the estimated change row.

**Comment by falnatsheh** [comment]
@jsom if you have time when you do acceptance for #158329465, can you try to see if you could better understand TableView inconsistent state that causes the error. So far what I think is that if you try to call `reloadRows` inside `tableView:cellForRowAt` sometimes the TableView gets into bad state, replacing `reloadRows` with updating the cell directly fixes the issue.

File: TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift [hunk 4]
```diff
@@ -416,12 +416,29 @@ class TradeItYahooEquityTradingTicketViewController: TradeItYahooViewController,
...
+                let errorResult = TradeItErrorResult(
+                    title: "Order Quantity Type Error",
+                    message: "No valid quantity type found for this action",
+                    code: TradeItErrorCode.systemError
+                )
+                self.alertManager.showError(errorResult, onViewController: self)
+                return cell
+            }
+            
+            self.order.quantityType = supportedOrderQuantityTypes.contains(self.order.quantityType) ? self.order.quantityType : supportedOrderQuantityTypes[0]
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift]: This code can still explode if someone accidentally removes the `isEmpty` check above. Better to use a guard?

```swift
guard let firstSupportedOrderQuantityType = supportedOrderQuantityTypes.first
    else {
        // show error
    }

self.order.quantityType = supportedOrderQuantityTypes.contains(self.order.quantityType) ? self.order.quantityType : firstSupportedOrderQuantityType
```

File: TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift [hunk 5]
```diff
@@ -529,6 +533,26 @@ class TradeItYahooEquityTradingTicketViewController: TradeItYahooViewController,
...
+    private func estimatedChangeText() -> String {
+        var estimateChangeText = "N/A"
+        
+        if let estimatedChange = order.estimatedChange() {
+            estimateChangeText = NumberFormatter.formatCurrency(
+                estimatedChange,
+                currencyCode: order.linkedBrokerAccount?.accountBaseCurrency
+            )
+        }
+        return estimateChangeText
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift]: I usually prefer a guard in cases like this to reduce nesting:
```swift
guard let estimatedChange = order.estimatedChange()
    else {
        return "N/A"
    }
return NumberFormatter.formatCurrency(
    estimatedChange,
    currencyCode: order.linkedBrokerAccount?.accountBaseCurrency
)
```

File: TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift (hunk 4 above)
Comment (falnatsheh) [reply TradeItIosTicketSDK2/TradeItYahooEquityTradingTicketViewController.swift]: Yeah the more swift-ey way of doing it

File: TradeItIosTicketSDK2/EquityPreviewCellFactory.swift [hunk 6]
```diff
@@ -72,6 +72,19 @@ class EquityPreviewCellFactory: PreviewCellFactory {
 
     // MARK: Private
 
+    private func formatQuantity(rawQuantityType: String, quantity: NSNumber) -> String{
```
Comment (falnatsheh) [inline TradeItIosTicketSDK2/EquityPreviewCellFactory.swift]: Ticket: https://www.pivotaltracker.com/story/show/157754041

Not very happy with the method but this is my first stab at it. Notes: 

- The `NumberFormatter.formatQuantity` defaults to 2, we could change/add "Quantity formatter" method that uses `orderCapabilities?.maxDecimalPlacesFor`.
- `orderDetails.orderQuantityType` is a string that in order to use it with `orderCapabilities?.maxDecimalPlacesFor` need to be convert to enum. If `EquityPreviewCellFactory.swift`  received the full `order` object or orderQuantityType enum the conversion will not be needed.

Comment (jsom) [reply TradeItIosTicketSDK2/EquityPreviewCellFactory.swift]: @falnatsheh yeah I think adding a new parameter to `NumberFormatter.formatQuantity` makes sense

**Comment by jsom** [comment]
@falnatsheh there also seems to be a WIP commit at the top here. Can you update it with a good commit message if you want to keep it?

**Pull request closed without merging**